### PR TITLE
rootless: fix a regression when using -d

### PR DESCRIPTION
**Pull request by giuseppe**
when using -d and port mapping, make sure the correct fd is injected
into conmon.

Move the pipe creation earlier as the fd must be known at the time we
create the container through conmon.

Closes: https://github.com/containers/libpod/issues/5167

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: libpod/oci_conmon_linux.go [hunk 1]
```diff
@@ -1160,6 +1160,14 @@ func (r *ConmonOCIRuntime) createOCIContainer(ctr *Container, restoreOptions *Co
...
 
 	if ctr.config.NetMode.IsSlirp4netns() {
+		havePortMapping := len(ctr.Config().PortMappings) > 0
+		if havePortMapping {
+			ctr.rootlessPortSyncR, ctr.rootlessPortSyncW, err = os.Pipe()
+			if err != nil {
+				return errors.Wrapf(err, "failed to create rootless port sync pipe")
+			}
+		}
+
```
Comment (TomSweeneyRedHat) [inline libpod/oci_conmon_linux.go]: should you do the `defer errorhandling` for rootlessPortSyncR/W here as you do below for rootlessSlipSyncR/W around line 1177?

File: libpod/networking_linux.go [hunk 2]
```diff
@@ -355,6 +351,11 @@ func (r *Runtime) setupRootlessPortMapping(ctr *Container, netnsPath string) (er
 	cmd := exec.Command(fmt.Sprintf("/proc/%d/exe", os.Getpid()))
 	cmd.Args = []string{rootlessport.ReexecKey}
 	// Leak one end of the pipe in rootlessport process, the other will be sent to conmon
+
+	if ctr.rootlessPortSyncR != nil {
+		defer errorhandling.CloseQuiet(ctr.rootlessPortSyncR)
+	}
+
```
Comment (TomSweeneyRedHat) [inline libpod/networking_linux.go]: Do you need a similar for rootlessSlirpSyncR?

File: libpod/oci_conmon_linux.go (hunk 1 above)
Comment (giuseppe) [reply libpod/oci_conmon_linux.go]: no, we cannot close `rootlessPortSyncR` here, it must be left open as it used later

File: libpod/networking_linux.go (hunk 2 above)
Comment (giuseppe) [reply libpod/networking_linux.go]: do you mean `rootlessSlirpSyncW`?  No, it is already closed in `libpod/oci_conmon_linux.go`

**Comment by rhatdan** [comment]
LGTM
but tests are not happy.

**Comment by giuseppe** [comment]
tests are passing now

**Pull request merged**